diff --git a/Edu-Android/testResources/META-INF/plugin.xml b/Edu-Android/testResources/META-INF/plugin.xml
index 42c32aa1a..9255cd7d4 100644
--- a/Edu-Android/testResources/META-INF/plugin.xml
+++ b/Edu-Android/testResources/META-INF/plugin.xml
@@ -1,5 +1,6 @@
 <idea-plugin xmlns:xi="http://www.w3.org/2001/XInclude">
     <id>com.jetbrains.edu.android</id>
+    <xi:include href="/META-INF/educational-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
     <depends optional="true" config-file="Edu-Android.xml">org.jetbrains.android</depends>

     <extensions defaultExtensionNs="Educational">
diff --git a/Edu-Java/testResources/META-INF/plugin.xml b/Edu-Java/testResources/META-INF/plugin.xml
index c09e7e73d..5278ba576 100644
--- a/Edu-Java/testResources/META-INF/plugin.xml
+++ b/Edu-Java/testResources/META-INF/plugin.xml
@@ -1,5 +1,6 @@
 <idea-plugin xmlns:xi="http://www.w3.org/2001/XInclude">
     <id>com.jetbrains.edu.java</id>
+    <xi:include href="/META-INF/educational-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
     <depends optional="true" config-file="Edu-Java.xml">com.intellij.modules.java</depends>

     <extensions defaultExtensionNs="Educational">
diff --git a/Edu-Kotlin/testResources/META-INF/plugin.xml b/Edu-Kotlin/testResources/META-INF/plugin.xml
index 1f4d26990..c9fb71c71 100644
--- a/Edu-Kotlin/testResources/META-INF/plugin.xml
+++ b/Edu-Kotlin/testResources/META-INF/plugin.xml
@@ -1,5 +1,6 @@
 <idea-plugin xmlns:xi="http://www.w3.org/2001/XInclude">
     <id>com.jetbrains.edu.kotlin</id>
+    <xi:include href="/META-INF/educational-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
     <depends optional="true" config-file="Edu-Kotlin.xml">com.intellij.modules.java</depends>

     <extensions defaultExtensionNs="Educational">
diff --git a/Edu-Python/src/com/jetbrains/edu/python/learning/newproject/PyCourseProjectGenerator.java b/Edu-Python/src/com/jetbrains/edu/python/learning/newproject/PyCourseProjectGenerator.java
index 47d00037d..d17ba9143 100644
--- a/Edu-Python/src/com/jetbrains/edu/python/learning/newproject/PyCourseProjectGenerator.java
+++ b/Edu-Python/src/com/jetbrains/edu/python/learning/newproject/PyCourseProjectGenerator.java
@@ -81,14 +81,14 @@ public class PyCourseProjectGenerator extends CourseProjectGenerator<PyNewProjec
           final PyPackageManager packageManager = PyPackageManager.getInstance(baseSdk);
           return packageManager.createVirtualEnv(virtualEnvPath, false);
         }
-      }, getAllSdks(), baseSdk, project.getBasePath(), null);
+      }, getAllSdks(), baseSdk, project.getBasePath());
       if (sdk == null) {
         LOG.warn("Failed to create virtual env in " + virtualEnvPath);
         return;
       }
       settings.setSdk(sdk);
       SdkConfigurationUtil.addSdk(sdk);
-      PySdkExtKt.associateWithModule(sdk, null, project.getBasePath());
+      PySdkExtKt.associateWithProject(sdk, project, false);
     }
   }

diff --git a/Edu-Python/src/com/jetbrains/edu/python/learning/pycharm/PyLanguageSettings.kt b/Edu-Python/src/com/jetbrains/edu/python/learning/pycharm/PyLanguageSettings.kt
index 3f1a87520..ca30893a7 100644
--- a/Edu-Python/src/com/jetbrains/edu/python/learning/pycharm/PyLanguageSettings.kt
+++ b/Edu-Python/src/com/jetbrains/edu/python/learning/pycharm/PyLanguageSettings.kt
@@ -17,7 +17,7 @@ internal open class PyLanguageSettings : PyLanguageSettings() {
       if (it != null && PythonSdkType.isVirtualEnv(it)) {
         val data = it.sdkAdditionalData as PythonSdkAdditionalData?
         if (data != null) {
-          val path = data.associatedModulePath
+          val path = data.associatedProjectPath
           if (path != null) {
             return@removeIf true
           }
@@ -27,7 +27,7 @@ internal open class PyLanguageSettings : PyLanguageSettings() {
     }

     val sdks = if (fakeSdk != null) ContainerUtil.prepend(registeredSdks, fakeSdk) else registeredSdks
-    val sdkChooser = PythonSdkChooserCombo(null, null, sdks, null) { true }
+    val sdkChooser = PythonSdkChooserCombo(null, sdks, null) { true }
     sdkChooser.addChangedListener {
       mySettings.sdk = sdkChooser.comboBox.selectedItem as? Sdk
       notifyListeners()
diff --git a/Edu-Python/testResources/META-INF/plugin.xml b/Edu-Python/testResources/META-INF/plugin.xml
index 5c7158398..a20004f8c 100644
--- a/Edu-Python/testResources/META-INF/plugin.xml
+++ b/Edu-Python/testResources/META-INF/plugin.xml
@@ -1,4 +1,5 @@
 <idea-plugin xmlns:xi="http://www.w3.org/2001/XInclude">
     <id>com.jetbrains.edu.python</id>
+    <xi:include href="/META-INF/educational-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
     <depends optional="true" config-file="Edu-Python.xml">com.intellij.modules.python</depends>
 </idea-plugin>
diff --git a/Edu-Scala/testResources/META-INF/plugin.xml b/Edu-Scala/testResources/META-INF/plugin.xml
index 5eca860a3..edce8141e 100644
--- a/Edu-Scala/testResources/META-INF/plugin.xml
+++ b/Edu-Scala/testResources/META-INF/plugin.xml
@@ -1,4 +1,5 @@
 <idea-plugin xmlns:xi="http://www.w3.org/2001/XInclude">
     <id>com.jetbrains.edu.scala</id>
+    <xi:include href="/META-INF/educational-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
     <depends optional="true" config-file="Edu-Scala.xml">org.intellij.scala</depends>
 </idea-plugin>
diff --git a/build.gradle b/build.gradle
index e3cb091cc..fc99977ab 100644
--- a/build.gradle
+++ b/build.gradle
@@ -154,7 +154,6 @@ allprojects {
     }

     dependencies {
-        compileOnly "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlinVersion"
         compile group: 'org.twitter4j', name: 'twitter4j-core', version: '4.0.1'
         compile 'org.jsoup:jsoup:1.11.2'
         compile("org.jetbrains:markdown:0.1.28") {
@@ -209,7 +208,7 @@ intellij {
     pluginName 'EduTools'
     updateSinceUntilBuild true
     downloadSources false
-    plugins 'junit', 'Kotlin', "PythonCore:$pythonPluginVersion", "org.intellij.scala:$scalaPluginVersion", "NodeJS:$nodeJsPluginVersion", "JavaScriptLanguage"
+    plugins 'junit', 'Kotlin', "PythonCore:$pythonPluginVersion", "org.intellij.scala:$scalaPluginVersion"
 }

 task configurePyCharm {
@@ -304,7 +303,7 @@ configure(project(':educational-core')) {
     }
 }

-configure([project(':Edu-Python'), project(':Edu-Java'), project(':Edu-Kotlin'), project(':Edu-Scala'), project(':Edu-Android'), project(':Edu-JavaScript')]) {
+configure([project(':Edu-Python'), project(':Edu-Java'), project(':Edu-Kotlin'), project(':Edu-Scala'), project(':Edu-Android')]) {
     dependencies {
         compile project(':educational-core')
         testCompile project(':educational-core').sourceSets.test.output
@@ -361,26 +360,27 @@ configure(project(':Edu-Android')) {
     }
 }

-configure(project(':Edu-JavaScript')) {
-  task createCheckiOOAuthProperties() {
-    createProperties(
-      "${projectDir}/resources/checkio/js-checkio-oauth.properties",
-      [
-        "jsCheckioClientId" : jsCheckioClientId,
-        "jsCheckioClientSecret" : jsCheckioClientSecret
-      ]
-    )
-  }
-  processResources.dependsOn(createCheckiOOAuthProperties)
-
-  intellij.plugins "NodeJS:$nodeJsPluginVersion", "JavaScriptLanguage"
-}
+//no JS support in Android Studio
+//configure(project(':Edu-JavaScript')) {
+//  task createCheckiOOAuthProperties() {
+//    createProperties(
+//      "${projectDir}/resources/checkio/js-checkio-oauth.properties",
+//      [
+//        "jsCheckioClientId" : jsCheckioClientId,
+//        "jsCheckioClientSecret" : jsCheckioClientSecret
+//      ]
+//    )
+//  }
+//  processResources.dependsOn(createCheckiOOAuthProperties)
+//
+//  intellij.plugins "NodeJS:$nodeJsPluginVersion", "JavaScriptLanguage"
+//}

 runIde.systemProperty("-Didea.is.internal", "true")
 runIde.systemProperty("-ea", "")

 dependencies {
-    compile project(':educational-core'), project(':Edu-Python'), project(':Edu-Kotlin'), project(':Edu-Java'), project(':Edu-Scala'), project(':Edu-Android'), project(':Edu-JavaScript')
+    compile project(':educational-core'), project(':Edu-Python'), project(':Edu-Kotlin'), project(':Edu-Java'), project(':Edu-Scala'), project(':Edu-Android')
 }

 idea {
diff --git a/educational-core/src/com/jetbrains/edu/learning/editor/EduSingleFileEditor.kt b/educational-core/src/com/jetbrains/edu/learning/editor/EduSingleFileEditor.kt
index 635fe18f3..770bf0615 100644
--- a/educational-core/src/com/jetbrains/edu/learning/editor/EduSingleFileEditor.kt
+++ b/educational-core/src/com/jetbrains/edu/learning/editor/EduSingleFileEditor.kt
@@ -43,9 +43,9 @@ class EduSingleFileEditor(
     return EduEditorState(state, null)
   }

-  override fun setState(state: FileEditorState, exactState: Boolean) {
+  override fun setState(state: FileEditorState) {
     val realState = (state as? EduEditorState)?.mainEditorState ?: state
-    super<PsiAwareTextEditorImpl>.setState(realState, exactState)
+    super<PsiAwareTextEditorImpl>.setState(realState)
   }

   override fun validateTaskFile() {
diff --git a/educational-core/src/com/jetbrains/edu/learning/editor/EduSplitEditor.kt b/educational-core/src/com/jetbrains/edu/learning/editor/EduSplitEditor.kt
index 8cbed839c..583a754c2 100644
--- a/educational-core/src/com/jetbrains/edu/learning/editor/EduSplitEditor.kt
+++ b/educational-core/src/com/jetbrains/edu/learning/editor/EduSplitEditor.kt
@@ -55,11 +55,11 @@ class EduSplitEditor(
     return EduEditorState(mainEditorState.mainEditorState, secondaryEditor.getState(level))
   }

-  override fun setState(state: FileEditorState, exactState: Boolean) {
+  override fun setState(state: FileEditorState) {
     if (state is EduEditorState) {
-      mainEditor.setState(state, exactState)
+      mainEditor.setState(state)
       state.secondaryEditorState?.also {
-        secondaryEditor.setState(it, exactState)
+        secondaryEditor.setState(it)
       }
     }
   }
diff --git a/educational-core/src/icons/EducationalCoreIcons.java b/educational-core/src/icons/EducationalCoreIcons.java
index fe572b2fa..e40220997 100644
--- a/educational-core/src/icons/EducationalCoreIcons.java
+++ b/educational-core/src/icons/EducationalCoreIcons.java
@@ -39,7 +39,7 @@ public class EducationalCoreIcons {

   public static final Icon CourseAction = load("/icons/com/jetbrains/edu/eduCourseAction.png"); // 16x16
   public static final Icon CourseTree = load("/icons/com/jetbrains/edu/eduCourseTree.png"); // 16x16
-  public static final Icon CourseToolWindow = load("/icons/com/jetbrains/edu/eduCourseTask.svg"); // 13x13
+  public static final Icon CourseToolWindow = load("/icons/com/jetbrains/edu/eduCourseTask181.png"); // 13x13

   public static final Icon ResultCorrect = load("/icons/com/jetbrains/edu/learning/resultCorrect.svg"); // 16x16
   public static final Icon ResetTask = load("/icons/com/jetbrains/edu/learning/resetTask.svg"); // 16x16
diff --git a/educational-core/testSrc/com/jetbrains/edu/learning/courseView/NodesTest.kt b/educational-core/testSrc/com/jetbrains/edu/learning/courseView/NodesTest.kt
index d05af4596..c4c78bd64 100644
--- a/educational-core/testSrc/com/jetbrains/edu/learning/courseView/NodesTest.kt
+++ b/educational-core/testSrc/com/jetbrains/edu/learning/courseView/NodesTest.kt
@@ -156,9 +156,9 @@ class NodesTest: CourseViewTestBase() {
            taskFile3.txt
           taskFile1.txt
          -TaskNode task2
-          additionalFile1.txt
           -DirectoryNode folder
            additionalFile3.txt
+          additionalFile1.txt
     """.trimIndent())
   }

@@ -176,11 +176,11 @@ class NodesTest: CourseViewTestBase() {
           taskFile1.txt
           CCStudentInvisibleFileNode taskFile2.txt
          -CCTaskNode task2
-          additionalFile1.txt
-          CCStudentInvisibleFileNode additionalFile2.txt
           -CCNode folder
            additionalFile3.txt
            CCStudentInvisibleFileNode additionalFile4.txt
+          additionalFile1.txt
+          CCStudentInvisibleFileNode additionalFile2.txt
           CCStudentInvisibleFileNode task.html
     """.trimIndent())
   }
@@ -206,12 +206,12 @@ class NodesTest: CourseViewTestBase() {
           taskFile1.txt
           CCStudentInvisibleFileNode taskFile2.txt
          -CCTaskNode task2
-          additionalFile1.txt
-          CCStudentInvisibleFileNode additionalFile2.txt
           -CCNode folder
            additionalFile3.txt
            CCStudentInvisibleFileNode additionalFile4.txt
            CCStudentInvisibleFileNode non_course_file3.txt (excluded)
+          additionalFile1.txt
+          CCStudentInvisibleFileNode additionalFile2.txt
           CCStudentInvisibleFileNode task.html
         CCStudentInvisibleFileNode non_course_file1.txt
     """.trimIndent())
diff --git a/resources/META-INF/plugin.xml b/resources/META-INF/plugin.xml
index ae3188cb2..7d7d4a1e5 100644
--- a/resources/META-INF/plugin.xml
+++ b/resources/META-INF/plugin.xml
@@ -8,6 +8,7 @@
     <!--update changes in changes.html file instead-->
     <change-notes/>
     <depends>com.intellij.modules.lang</depends>
+    <depends>com.intellij.modules.androidstudio</depends>

     <xi:include href="/META-INF/educational-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
     <depends optional="true" config-file="Edu-Scala.xml">org.intellij.scala</depends>
diff --git a/settings.gradle b/settings.gradle
index b120c4d1e..1d02a34bc 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -1,2 +1,2 @@
 rootProject.name = 'educational-plugin'
-include 'Edu-Java', 'Edu-Scala', 'Edu-Kotlin', 'Edu-Python', 'educational-core', 'Edu-Android', 'Edu-JavaScript'
\ No newline at end of file
+include 'Edu-Java', 'Edu-Scala', 'Edu-Kotlin', 'Edu-Python', 'educational-core', 'Edu-Android'
\ No newline at end of file
