diff --git a/build.gradle b/build.gradle
index 3d9c578..9bc2606 100644
--- a/build.gradle
+++ b/build.gradle
@@ -205,7 +205,7 @@ intellij {
     updateSinceUntilBuild true
     downloadSources false
     plugins 'junit', 'Kotlin', "PythonCore:$pythonPluginVersion", "org.intellij.scala:$scalaPluginVersion", 
-            "NodeJS:$nodeJsPluginVersion", "JavaScriptLanguage", "org.rust.lang:$rustPluginVersion"
+            "org.rust.lang:$rustPluginVersion"
 }
 
 task configurePyCharm {
@@ -288,7 +288,6 @@ configure([project(':Edu-Python'),
            project(':Edu-Kotlin'), 
            project(':Edu-Scala'), 
            project(':Edu-Android'), 
-           project(':Edu-JavaScript'),
            project(':Edu-Rust')]) {
     dependencies {
         compile project(':educational-core')
@@ -324,9 +323,10 @@ configure(project(':Edu-Android')) {
     }
 }
 
-configure(project(':Edu-JavaScript')) {
-  intellij.plugins "NodeJS:$nodeJsPluginVersion", "JavaScriptLanguage"
-}
+//no JS support in Android Studio
+//configure(project(':Edu-JavaScript')) {
+//  intellij.plugins "NodeJS:$nodeJsPluginVersion", "JavaScriptLanguage"
+//}
 
 configure(project(':Edu-Rust')) {
     intellij.plugins "org.rust.lang:$rustPluginVersion"  
@@ -337,7 +337,7 @@ runIde.systemProperty("-ea", "")
 
 dependencies {
     compile project(':educational-core'), project(':Edu-Python'), project(':Edu-Kotlin'), project(':Edu-Java'), 
-            project(':Edu-Scala'), project(':Edu-Android'), project(':Edu-JavaScript'), project(':Edu-Rust')
+            project(':Edu-Scala'), project(':Edu-Android'), project(':Edu-Rust')
 }
 
 idea {
diff --git a/gradle.properties b/gradle.properties
index b934980..3f96688 100644
--- a/gradle.properties
+++ b/gradle.properties
@@ -1,4 +1,4 @@
-environmentName=183
+environmentName=studio-183
 
 stepikTestClientId=
 stepikTestClientSecret=
diff --git a/resources/META-INF/plugin.xml b/resources/META-INF/plugin.xml
index 96161c9..2ccd4c1 100644
--- a/resources/META-INF/plugin.xml
+++ b/resources/META-INF/plugin.xml
@@ -8,6 +8,7 @@
     <!--update changes in changes.html file instead-->
     <change-notes/>
     <depends>com.intellij.modules.lang</depends>
+    <depends>com.intellij.modules.androidstudio</depends>
 
     <xi:include href="/META-INF/educational-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
     <depends optional="true" config-file="Edu-Scala.xml">org.intellij.scala</depends>
diff --git a/settings.gradle b/settings.gradle
index 5c663d2..5a81daa 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -1,2 +1,2 @@
 rootProject.name = 'educational-plugin'
-include 'Edu-Java', 'Edu-Scala', 'Edu-Kotlin', 'Edu-Python', 'educational-core', 'Edu-Android', 'Edu-JavaScript', 'Edu-Rust'
\ No newline at end of file
+include 'Edu-Java', 'Edu-Scala', 'Edu-Kotlin', 'Edu-Python', 'educational-core', 'Edu-Android', 'Edu-Rust'
